core/modules/catalog/store/stock/actions.ts

```diff
@@ -1,92 +1,150 @@
-import StockState from '../../types/StockState'
-import rootStore from '@vue-storefront/core/store'
-import { TaskQueue } from '@vue-storefront/core/lib/sync'
-import { Logger } from '@vue-storefront/core/lib/logger'
-import config from 'config'
+import * as types from '@vue-storefront/core/modules/cart/store/mutation-types';
+import RootState from '@vue-storefront/core/types/RootState';
+import StockState from '../../types/StockState';
+import rootStore from '@vue-storefront/core/store';
+import { TaskQueue } from '@vue-storefront/core/lib/sync';
+import { Logger } from '@vue-storefront/core/lib/logger';
+import config from 'config';
 
 const actions: ActionTree<StockState, RootState> = {
   /**
    * Reset current configuration and selected variatnts
    */
-  check (context, { product, qty = 1 }) {
-    return new Promise((resolve, reject) => {
-      if (config.stock.synchronize) {
-        TaskQueue.queue({ url: config.stock.endpoint + '/check?sku=' + encodeURIComponent(product.sku),
-          payload: {
-            method: 'GET',
-            headers: { 'Content-Type': 'application/json' },
-            mode: 'cors'
-          },
-          product_sku: product.sku,
-          callback_event: 'store:stock/stockAfterCheck'
-        }).then((task: any) => {
-          resolve({ qty: product.stock ? product.stock.qty : 0, status: product.stock ? (product.stock.is_in_stock ? 'ok' : 'out_of_stock') : 'ok', onlineCheckTaskId: task.task_id }) // if online we can return ok because it will be verified anyway
-        })
-      } else {
-        resolve({ qty: product.stock ? product.stock.qty : 0, status: product.stock ? (product.stock.is_in_stock ? 'ok' : 'out_of_stock') : 'volatile' }) // if not online, cannot check the source of true here
-      }
-    })
+  async check (context, { product, qty = 1 }) {
+    if (rootStore.state.config.stock.synchronize) {
+      const task: any = await TaskQueue.execute({
+        url:
+          rootStore.state.config.stock.endpoint +
+          '/check?sku=' +
+          encodeURIComponent(product.sku),
+        payload: {
+          method: 'GET',
+          headers: { 'Content-Type': 'application/json' },
+          mode: 'cors'
+        },
+        product_sku: product.sku,
+        callback_event: 'store:stock/stockAfterCheck'
+      });
+      let result = task.result;
+      const cartProduct = await rootStore.dispatch('cart/getItem', product.sku);
+      let totalQty = cartProduct ? cartProduct.qty + product.qty : product.qty;
+      const status = result.is_in_stock
+        ? totalQty > result.qty
+          ? 'not_enough_stock'
+          : 'ok'
+        : 'out_of_stock';
+      return Object.assign({}, task, {
+        qty: task.result.qty,
+        status,
+        onlineCheckTaskId: task.task_id
+      });
+    } else {
+      return {
+        qty: product.stock ? product.stock.qty : 0,
+        status: product.stock
+          ? product.stock.is_in_stock
+            ? 'ok'
```

Comment:
could we use consts defined at the top of the file for these statuses?

Reply:
It seems to me that this will not improve readability in this case. Unless I do not see something;)